channels_redis/alt.py

```diff
@@ -0,0 +1,311 @@
+        return self._shards[shard_index]
+
+    def _get_group_channel_name(self, group):
+        """
+        Return the channel name used by a group.
+        Includes '__group__' in the returned
+        string so that these names are distinguished
+        from those returned by `new_channel()`.
+        Technically collisions are possible, but it
+        takes what I believe is intentional abuse in
+        order to have colliding names.
+        """
+        return f"{self.prefix}__group__{group}"
+
+    extensions = ["groups"]
+
+    ################################################################################
+    # Channel layer API
+    ################################################################################
+
+    async def send(self, channel, message):
+        """
+        Send a message onto a (general or specific) channel.
+        """
+        shard = self._get_shard(channel)
+        await shard.publish(channel, message)
+
+    async def new_channel(self, prefix="specific."):
+        """
+        Returns a new channel name that can be used by a consumer in our
+        process as a specific channel.
+        """
+        channel = f"{self.prefix}{prefix}{uuid.uuid4().hex}"
+        self.channels[channel] = asyncio.Queue()
+        shard = self._get_shard(channel)
+        await shard.subscribe(channel)
+        return channel
+
+    async def receive(self, channel):
+        """
+        Receive the first message that arrives on the channel.
+        If more than one coroutine waits on the same channel, a random one
+        of the waiting coroutines will get the result.
+        """
+        if channel not in self.channels:
+            raise RuntimeError(
+                'You should only call receive() on channels that you "own" and that were created with `new_channel()`.'
+            )
+
+        q = self.channels[channel]
+
+        try:
+            message = await q.get()
+        except asyncio.CancelledError:
+            # We assume here that the reason we are cancelled is because the consumer
+            # is exiting, therefore we need to cleanup by unsubscribe below. Indeed,
+            # currently the way that Django Channels works, this is a safe assumption.
+            # In the future, Dajngo Channels could change to call a *new* method that
+            # would serve as the antithesis of `new_channel()`; this new method might
+            # be named `delete_channel()`. If that were the case, we would do the
+            # following cleanup from that new `delete_channel()` method, but, since
+            # that's not how Django Channels works (yet), we do the cleanup below:
+            if channel in self.channels:
+                del self.channels[channel]
+                try:
+                    shard = self._get_shard(channel)
+                    await shard.unsubscribe(channel)
+                except Exception:
+                    traceback.print_exc(file=sys.stderr)
```

Comment:
wouldnt be better to use `logging.exception()` here?

Reply:
Yes, I agree. Although we don't have access to an instantiated `Logger` object here. We could create our own logger (as in `logging.getLogger(__name__)`) but I don't like that either. It would be best to hook into an existing logger (like Uvicorn and/or Daphne each instantiate a logger, it would be great to have access to _it_ here). Thoughts?

Reply:
> We could create our own logger (as in logging.getLogger(__name__))

Sounds good to me. It's done the same way in here in `channels_redis/core.py`, similarly in Channels themselves, so it would be nicely consistent. 

Plus in case someone is using Sentry etc., this exception wouldn't be collected as it is now.

Reply:
Okay I'm convinced. I'll do it like `channels_redis/core.py` does.